File: Ghpr.LocalFileSystem/Mappers/TestRunDtoMapper.cs
using System.Linq;
using Ghpr.Core.Common;
using Ghpr.LocalFileSystem.Entities;

namespace Ghpr.LocalFileSystem.Mappers
{
    public static class TestRunDtoMapper
    {
        public static TestRun Map(this TestRunDto testRunDto, SimpleItemInfoDto testOutputInfoDto)
        {
            var testRun = new TestRun
            {
                Categories = testRunDto.Categories,
                Description = testRunDto.Description,
                Events = testRunDto.Events.Select(teDto => new TestEvent
                {
                    Started = teDto.Started,
                    Finished = teDto.Finished,
                    Comment = teDto.Comment,
                    EventInfo = teDto.EventInfo.MapSimpleItemInfo()
                }).ToList(),
                FullName = testRunDto.FullName,
                Name = testRunDto.Name,
                Output = testOutputInfoDto.MapSimpleItemInfo(),
                Priority = testRunDto.Priority,
                Result = testRunDto.Result,
                RunGuid = testRunDto.RunGuid,
                Screenshots = testRunDto.Screenshots.Select(sDto => sDto.MapSimpleItemInfo()).ToList(),
                TestInfo = testRunDto.TestInfo.MapTestRunInfo(),
                Duration = testRunDto.Duration.Equals(0.0)
                    ? (testRunDto.TestInfo.Finish - testRunDto.TestInfo.Start).TotalSeconds
                    : testRunDto.Duration,
                TestMessage = testRunDto.TestMessage,
                TestStackTrace = testRunDto.TestStackTrace,
                TestType = testRunDto.TestType,
                TestData = testRunDto.TestData.Select(tdDto => new TestData
                {
                    Actual = tdDto.Actual,
                    Expected = tdDto.Expected,
                    Comment = tdDto.Comment,
                    TestDataInfo = tdDto.TestDataInfo.MapSimpleItemInfo()
                }).ToList()
            };
            return testRun;
        }

        public static TestRunDto ToDto(this TestRun testRun, SimpleItemInfo testOutputInfo)
        {
            var testRunDto = new TestRunDto
            {
                Categories = testRun.Categories,
                Description = testRun.Description,
                Events = testRun.Events.Select(teDto => new TestEventDto
                {
                    Started = teDto.Started,
                    Finished = teDto.Finished,
                    Comment = teDto.Comment,
                    EventInfo = teDto.EventInfo.ToDto()
                }).ToList(),
                FullName = testRun.FullName,
                Name = testRun.Name,
                Output = testOutputInfo.ToDto(),
                Priority = testRun.Priority,
                Result = testRun.Result,
                RunGuid = testRun.RunGuid,
                Screenshots = testRun.Screenshots.Select(s => s.ToDto()).ToList(),
                TestInfo = testRun.TestInfo.ToDto(),
                Duration = testRun.Duration.Equals(0.0)
                    ? (testRun.TestInfo.Finish - testRun.TestInfo.Start).TotalSeconds
                    : testRun.Duration,
                TestMessage = testRun.TestMessage,
                TestStackTrace = testRun.TestStackTrace,
                TestType = testRun.TestType,
                TestData = testRun.TestData.Select(tdDto => new TestDataDto
                {
                    Actual = tdDto.Actual,
                    Expected = tdDto.Expected,
                    Comment = tdDto.Comment,
                    TestDataInfo = tdDto.TestDataInfo.ToDto()
                }).ToList()
            };
            return testRunDto;
        }
    }
}

File: Ghpr.Controller.ts/controller/common/localFileSystem/mappers/TestRunDtoMapper.ts
///<reference path="./../entities/TestRun.ts"/>
///<reference path="./../../../dto/TestRunDto.ts"/>
///<reference path="./ItemInfoDtoMapper.ts"/>
///<reference path="./SimpleItemInfoDtoMapper.ts"/>

class TestRunDtoMapper {
    static map(testRun: TestRun): TestRunDto {
        let eventDtos = new Array<TestEventDto>(testRun.events.length);
        let screenshotDtos = new Array<SimpleItemInfoDto>(testRun.screenshots.length);
        let testDataDtos = new Array<TestDataDto>(testRun.testData.length);

        for (let i = 0; i < testRun.events.length; i++) {
            let event = testRun.events[i];
            let eventDto = new TestEventDto();
            eventDto.name = event.name;
            eventDto.started = event.started;
            eventDto.finished = event.finished;
            eventDto.eventInfo = SimpleItemInfoDtoMapper.map(event.eventInfo);
            eventDtos[i] = eventDto;
        }

        for (let i = 0; i < testRun.screenshots.length; i++) {
            screenshotDtos[i] = SimpleItemInfoDtoMapper.map(testRun.screenshots[i]);
        }

        for (let i = 0; i < testRun.testData.length; i++) {
            let testData = testRun.testData[i];
            let testDataDto = new TestDataDto();
            testDataDto.actual = testData.actual;
            testDataDto.expected = testData.expected;
            testDataDto.comment = testData.comment;
            testDataDto.testDataInfo = SimpleItemInfoDtoMapper.map(testData.testDataInfo);
            testDataDtos[i] = testDataDto;
        }

        let testRunDto = new TestRunDto();
        testRunDto.name = testRun.name;
        testRunDto.categories = testRun.categories;
        testRunDto.description = testRun.description;
        if (testRun.duration !== undefined && testRun.duration !== 0) {
            testRunDto.duration = testRun.duration;
        } else {
            const diffTime = testRun.testInfo.finish.valueOf() - testRun.testInfo.start.valueOf();
            const diffSecs = Math.round(((diffTime / 1000) + Number.EPSILON) * 100) / 100;
            testRunDto.duration = diffSecs;
        }
        testRunDto.events = eventDtos;
        testRunDto.fullName = testRun.fullName;
        testRunDto.output = SimpleItemInfoDtoMapper.map(testRun.output);
        testRunDto.priority = testRun.priority;
        testRunDto.result = testRun.result;
        testRunDto.testInfo = ItemInfoDtoMapper.map(testRun.testInfo);
        testRunDto.testMessage = testRun.testMessage;
        testRunDto.testStackTrace = testRun.testStackTrace;
        testRunDto.runGuid = testRun.runGuid;
        testRunDto.testType = testRun.testType;
        testRunDto.screenshots = screenshotDtos;
        testRunDto.testData = testDataDtos;
        
        return testRunDto;
    }
}
